add gce_template.py

GwenaelPellenArkeup opened the pull request:
gce template manage GCE Instance Templates in Google Cloud Plateform.

ISSUE TYPE

New Module Pull Request

COMPONENT NAME

gce_template

ANSIBLE VERSION

ansible 2.3

SUMMARY

Module can create and detroy Intance Templates.

GwenaelPellenArkeup committed:
add gce_template.py

gce template manage GCE Instance Templates in Google Cloud Plateform.

GwenaelPellenArkeup commented:
sorry everybody, i make a cleaner pull-request.

supertom commented on lib/ansible/modules/cloud/google/gce_template.py:
```diff
@@ -0,0 +1,596 @@
+
+
+def get_info(inst):
+    """Retrieves instance template informations
```
No 's' on information.

supertom commented on lib/ansible/modules/cloud/google/gce_template.py:
```diff
@@ -0,0 +1,596 @@
+
+
+def create_instance_template(module, gce):
+    """Create on instance template
```
"an instance template", not "on instance template".

supertom commented on lib/ansible/modules/cloud/google/gce_template.py:
```diff
@@ -0,0 +1,596 @@
+
+    Returns:
+        instance template information
+
```
delete blank line here.

supertom commented on lib/ansible/modules/cloud/google/gce_template.py:
```diff
@@ -0,0 +1,596 @@
+
+EXAMPLES = '''
+# Usage
+
```
Please delete blank line

supertom commented on lib/ansible/modules/cloud/google/gce_template.py:
```diff
@@ -0,0 +1,596 @@
+EXAMPLES = '''
+# Usage
+
+    gce_template:
```
please pull gce_template to the left margin.

supertom commented on lib/ansible/modules/cloud/google/gce_template.py:
```diff
@@ -0,0 +1,596 @@
+    choices: [
+      "bigquery", "cloud-platform", "compute-ro", "compute-rw",
+      "useraccounts-ro", "useraccounts-rw", "datastore", "logging-write",
+      "monitoring", "sql", "sql-admin", "storage-full", "storage-ro",
```
remove 'sql' from the choices list.  It will be removed from gce.py soon.

GwenaelPellenArkeup replied:
remove 'sql-admin' too ?

supertom commented on lib/ansible/modules/cloud/google/gce_template.py:
```diff
@@ -0,0 +1,596 @@
+
+    Returns:
+        instance template information
+
```
please delete blank line here.

supertom commented on lib/ansible/modules/cloud/google/gce_template.py:
```diff
@@ -0,0 +1,596 @@
+    if instance:
+        json_data = get_info(instance)
+    else:
+        module.fail_json(msg="no instance template !")
```
delete space between template and '!'

supertom commented on lib/ansible/modules/cloud/google/gce_template.py:
```diff
@@ -0,0 +1,596 @@
+      >= 0.20.0 if using preemptible option"
+notes:
+  - JSON credentials strongly preferred.
+author: "Gwenael Pellen <gwenael.pellen@arkeup.com>"
```
Please include your github ID in the author line, in between your name and email address.  For example:

`"my name (@my-github) <my-email@example.com>"`

supertom reviewed (changes requested):
Thanks for the PR.  Just some changes based on style/formatting.

supertom commented:
@GwenaelPellenArkeup thanks for submitting.  I tested it out and it works as expected.  I'm OK with letting it in as-is provided the style/formatting changes are made.  Going forward, someone will need to consolidate the duplicated code between [gce.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/google/gce.py) and this one, but, I don't want to keep the functionality from being included due to housekeeping.

GwenaelPellenArkeup committed:
change gce_template on style/formating

Apply change on style/formating from reviewer request.

GwenaelPellenArkeup committed:
change gce_template on style/formating again.

supertom commented:
@GwenaelPellenArkeup thanks for the changes.  re-tested and LGTM!  @ryansb I'm OK with merging this.

supertom commented:
Hi @GwenaelPellenArkeup, in talking with @ryansb, we should rename this gce_instance_template to be more specific.  Sound good?

GwenaelPellenArkeup committed:
Rename gce_template.py to gce_instance_template.py

GwenaelPellenArkeup committed:
update gce_instance_template.py to pass CI

Oops :)
ERROR: Import found before documentation variables. 
All imports must appear below DOCUMENTATION/EXAMPLES/RETURN/ANSIBLE_METADATA.

supertom reviewed (changes requested):
Thanks for the rename.  Would you please update the DOCUMENTATION and EXAMPLE blocks as well?  I see a few references to gce_template rather than gce_instance_template.

Thanks!

GwenaelPellenArkeup committed:
Update gce_instance_template.py

change documentation replace gce_template by gce_instance_template

GwenaelPellenArkeup committed:
Update gce_instance_template.py

Sorry for the typography fault. The day begins badly.

ryansb commented on lib/ansible/modules/cloud/google/gce_instance_template.py:
```diff
@@ -0,0 +1,582 @@
+short_description: create or destroy intance templates of Compute Engine of GCP.
+description:
+     - Creates or destroy Google instance templates
+       of Compte Engine of Google Cloud Plateform.
```
s/Compte/Compute/

s/Plateform/Platform/

ryansb commented on lib/ansible/modules/cloud/google/gce_instance_template.py:
```diff
@@ -0,0 +1,582 @@
+            state=dict(choices=['present', 'absent'], default='present'),
+            name=dict(require=True, aliases=['base_name']),
+            size=dict(default='f1-micro'),
+            source=dict(),
```
In the docs you mention that source and image cannot both be applied. Can you enforce this by adding an argument like this: `mutually_exclusive=[['source', 'image']],` around line 555?

ryansb commented on lib/ansible/modules/cloud/google/gce_instance_template.py:
```diff
@@ -0,0 +1,582 @@
+  image_family:
+    description:
+      - The image family to use to create the instance.
+        Cannot specify both 'image' and 'source'.
```
So image_family and image_source are required together, but neither can be specified if source is specified? Is that right?

GwenaelPellenArkeup replied:
If the souce is specified neither image nor image_family should be used.
Otherwise image or image_family must be specified.

ryansb commented on lib/ansible/modules/cloud/google/gce_instance_template.py:
```diff
@@ -0,0 +1,582 @@
+            source=dict(),
+            image=dict(),
+            image_family=dict(default='debian-8'),
+            disk_type=dict(default='pd-standard'),
```
Use an option type here, to allow only valid disk types (I guess `pd-ssd` and `pd-standard`?). @supertom are there more options?

ryansb commented on lib/ansible/modules/cloud/google/gce_instance_template.py:
```diff
@@ -0,0 +1,582 @@
+    default: "default"
+  subnetwork:
+    description:
+      - The Subnetwork resource for this instance.
```
What is the format for this field? Is it an ID, name, or something else?

ryansb commented on lib/ansible/modules/cloud/google/gce_instance_template.py:
```diff
@@ -0,0 +1,582 @@
+      - The Subnetwork resource for this instance.
+    required: false
+    default: null
+  can_ip_forword:
```
s/forword/forward/

GwenaelPellenArkeup committed:
Update gce_instance_template.py

apply text change
  s/Compte/Compute/
  s/Plateform/Platform/
  s/forword/forward/
change documentation for subnetwork, subnetwork is name.
add mutually_exclusive in AnsibleModule arguments.
change disk_type as an option type.

ryansb commented:
shipit

bjolivot commented:
ok after test
shipit

gundalow commented on lib/ansible/modules/cloud/google/gce_instance_template.py:
```diff
@@ -0,0 +1,583 @@
+options:
+  state:
+    description:
+      - desired state of instance template
```
Descriptions need to be full sentences, Capital letters & full stops

Please see docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html for more detail

gundalow commented on lib/ansible/modules/cloud/google/gce_instance_template.py:
```diff
@@ -0,0 +1,583 @@
+    default: null
+  size:
+    description:
+      - instance template size to use, use 'f1-micro' by default
```
Don't document default, that's what the `default:` field is for docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html

gundalow commented on lib/ansible/modules/cloud/google/gce_instance_template.py:
```diff
@@ -0,0 +1,583 @@
+  size:
+    description:
+      - instance template size to use, use 'f1-micro' by default
+    required: false
```
`required: false` files can be removed, see docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html for more info

gundalow commented on lib/ansible/modules/cloud/google/gce_instance_template.py:
```diff
@@ -0,0 +1,583 @@
+  source:
+    description:
+      - A source disk to attach to the instance.
+        Cannot specify both 'image' and 'source'.
```
`Cannot specify both I(image) and I(source).`
http://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html#formatting-options

gundalow commented on lib/ansible/modules/cloud/google/gce_instance_template.py:
```diff
@@ -0,0 +1,583 @@
+  image:
+    description:
+      - The image to use to create the instance.
+        Cannot specify both 'image' and 'source'.
```
`Cannot specify both I(image) and I(source).`
http://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html#formatting-options

gundalow commented on lib/ansible/modules/cloud/google/gce_instance_template.py:
```diff
@@ -0,0 +1,583 @@
+  image_family:
+    description:
+      - The image family to use to create the instance.
+        Cannot specify both 'image' and 'source'.
```
The doesn't appear to match the code
```
 +        mutually_exclusive=[['source', 'image']],
 +        required_one_of=[['image', 'image_family']],
 +        supports_check_mode=True
```

Please also use `I(...)` for option names

gundalow commented on lib/ansible/modules/cloud/google/gce_instance_template.py:
```diff
@@ -0,0 +1,583 @@
+    description:
+      - Indicate that the boot disk should be
+        deleted when the Node is deleted. Set to
+        True by default.
```
Remove default from `description:`

gundalow commented on lib/ansible/modules/cloud/google/gce_instance_template.py:
```diff
@@ -0,0 +1,583 @@
+  external_ip:
+    description:
+      - The external IP address to use.
+        If 'ephemeral' (default), a new non-static address will be
```
I(ephemeral)

gundalow commented on lib/ansible/modules/cloud/google/gce_instance_template.py:
```diff
@@ -0,0 +1,583 @@
+    description:
+      - The external IP address to use.
+        If 'ephemeral' (default), a new non-static address will be
+        used.  If 'None', then no external address will
```
I(None)

gundalow commented on lib/ansible/modules/cloud/google/gce_instance_template.py:
```diff
@@ -0,0 +1,583 @@
+    default: null
+  tags:
+    description:
+      - a comma-separated list of tags to associate with the instance
```
Might be better to define this as `type='list'`

gundalow commented on lib/ansible/modules/cloud/google/gce_instance_template.py:
```diff
@@ -0,0 +1,583 @@
+
+EXAMPLES = '''
+# Usage
+  name: my-test-instance-template
```
Missing `-` before `name:`

gundalow commented on lib/ansible/modules/cloud/google/gce_instance_template.py:
```diff
@@ -0,0 +1,583 @@
+author: "Gwenael Pellen (@GwenaelPellenArkeup) <gwenael.pellen@arkeup.com>"
+'''
+
+EXAMPLES = '''
```
Please review http://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html#examples-block

GwenaelPellenArkeup committed:
Update gce_instance_template.py

change the documentation.

GwenaelPellenArkeup committed:
Update gce_instance_template.py

RETURN is required for all new modules. Is empty because no new return variable returned.

GwenaelPellenArkeup commented:
bot_status

GwenaelPellenArkeup commented:
shipit

ryansb commented on lib/ansible/modules/cloud/google/gce_instance_template.py:
```diff
@@ -0,0 +1,565 @@
+        service_account_email: "{{ service_account_email }}"
+'''
+
+RETURN = '''
```
Please add docs on what this module returns.

GwenaelPellenArkeup replied:
Is empty because no new return variable returned.
Only standard return variable used.

ryansb commented on lib/ansible/modules/cloud/google/gce_instance_template.py:
```diff
@@ -0,0 +1,565 @@
+        json_output['changed'] = changed
+        json_output['msg'] = output
+    else:
+        module.fail_json(
```
This `else` is not needed, Ansible will verify that only present/absent are passed as the state.

GwenaelPellenArkeup committed:
Update gce_instance_template.py

Ansible will verify that only present/absent are passed as the state. This else is not needed.
